<div style="text-align: center; background-color: #750E21; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  FINAL PROJECT: RESEARCHING ON MUSIC TASTE WORDWIDELY 📌
</div>

<div style="text-align: center; background-color: #0766AD; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Stage 02 - Preprocessing 📌
</div>

## **IMPORT LIBRARY** 🎄

In [1]:
import numpy as np
import pandas as pd
from datetime import datetime

## **OVERAL INSIGHT INTO DATA** 📝

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  1. Read raw data 🔥
</div>

In [2]:
raw_df = pd.read_csv('../data/raw/final_raw_data.csv')
raw_df.head()

Id          View        Like Duration     Channel_name  \
0  9bZkp7q19f0  4.977100e+09  27830052.0  0:04:13      officialpsy   
1  hT_nvWreIhg  3.927344e+09  17491157.0  0:04:44  OneRepublicVEVO   
2  JGwWNGJdvx8  6.149237e+09  32323841.0  0:04:24       Ed Sheeran   
3  lp-EO5I60KA  3.700048e+09  14912926.0  0:04:57       Ed Sheeran   
4  CevxZvSJLk8  3.918655e+09  16596026.0  0:04:30    KatyPerryVEVO   

   Subscriber Country         Publish_time  Hasgtag  \
0  18400000.0     NaN  2012-07-15 07:46:32      4.0   
1   5470000.0     NaN  2013-05-31 07:00:36      2.0   
2  53900000.0     NaN  2017-01-30 10:57:50      3.0   
3  53900000.0     NaN  2014-10-07 13:57:37      3.0   
4  24600000.0      US  2013-09-05 20:00:22      0.0   

                                               Title  Most view per day  \
0                     PSY - GANGNAM STYLE(강남스타일) M/V           14924298   
1                       OneRepublic - Counting Stars            3288973   
2   Ed Sheeran - Shape of You (Official Music Video)           14390704   
3  Ed Sheeran - Thinking Out Loud (Official Music...            3771622   
4                                  Katy Perry - Roar           11294380   

  Most-view-date  Highest rank  Time to Highest rank  Charted-duration  \
0     2012/12/21           1.0                  36.0             482.0   
1     2018/11/10           4.0                   NaN             482.0   
2     2017/05/13           1.0                   4.0             356.0   
3     2015/02/14           3.0                   1.0             304.0   
4     2013/09/07           2.0                   5.0             445.0   

               Song Name  Artist Name Release Date (Spotify)  \
0  Gangnam Style (강남스타일)          PSY             2012-01-01   
1         Counting Stars  OneRepublic             2013-01-01   
2           Shape of You   Ed Sheeran             2017-03-03   
3      Thinking out Loud   Ed Sheeran             2014-06-21   
4                   Roar   Katy Perry             2013-10-18   

   Popularity Score (Spotify)  \
0                          75   
1                          88   
2                          88   
3                          85   
4                          74   

                                               Genre  
0             ['k-rap', 'korean old school hip hop']  
1                                 ['panamanian pop']  
2  ['colombian pop', 'dance pop', 'latin pop', 'p...  
3                  ['latin pop', 'puerto rican pop']  
4                            ['canadian pop', 'pop']

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  2. Comment on statistic of raw_df 🔥
</div>

In [3]:
raw_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Id                          2500 non-null   object 
 1   View                        2491 non-null   float64
 2   Like                        2488 non-null   float64
 3   Duration                    2491 non-null   object 
 4   Channel_name                2491 non-null   object 
 5   Subscriber                  2491 non-null   float64
 6   Country                     1384 non-null   object 
 7   Publish_time                2491 non-null   object 
 8   Hasgtag                     2491 non-null   float64
 9   Title                       2500 non-null   object 
 10  Most view per day           2500 non-null   int64  
 11  Most-view-date              2500 non-null   object 
 12  Highest rank                1876 non-null   float64
 13  Time to Highest rank        319 n

**Comment**
- Data Overview: The DataFrame has 2500 entries. There are 20 columns in total, with a mix of data types: float64, int64, and object. 
- dtypes: float64(7), int64(2), object(11)
- Columns contain full data (2500 data, reached 100%) including 9 columns: `Id`, `Title`, `Most view per day`, `Most-view-date`, `Song Name`, `Artist Name`, `Release Date (Spotify)`, `Popularity Score (Spotify)`, and `Genre`. 
- Columns contain data with less loss, just about 0.36% (2488-2491 data, reached 99.52% - 99.64%) including 7 columns: `View`, `Like`, `Duration`, `Channel_name`, `Subscriber`, `Publish_time`, and `Hasgtag`. Because this data is collected directly from Youtube, it accidentally collects videos that are currently in unlisted mode, so some of that data should be exploited, but this omission is not possible. significantly.
- Those colums have a significant number of missing values (non-null count is less than 2500) should not be used routinely for large-scale data analysis: `Country` (1384, reached 55.36%), `Highest rank` (1876, reached 75.04%), and `Charted-duration` (1876, reached 75.04%).
- This colum has a Supermajority number of missing values, need to drop: `Time to Highest rank` (319, reached 12.76%)
- Columns such as "Publish_time," "Most-view-date," and "Release Date (Spotify)" seem to contain date or time-related information.
- Because of missing data, numeric data columns are defaulted to float64 type, to match the real number type of NaN to store numeric data cells with missing data.

## **LET'S PROCESS** 📝

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  1. How many rows and how many columns does the raw data have? 🧮
</div>

👁️ First, we need to check the number of rows and columns to see if our data has at least 5 fields and 1000 observations.

In [4]:
shape = raw_df.shape

In [5]:
print(f"Current shape: {shape}")

if ((shape[0] >= 1000) & (shape[1] >= 5)):
    print(f"Our data is good!.")
else:
    print(f"May be we need to crawl data again.!")

Current shape: (2500, 20)
Our data is good!.


<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  2. What does each line mean? Does it matter if the lines have different meanings? 🧮
</div>

🔍🔍🔍 Understand the meaning of each line gives us a more overall understanding of the data.\
🔍🔍🔍 What our data is talking about or if some lines of data have different meanings, it will lead to the unconsistency of data and giving wrong insights.

**Comment** 
- Each line of data contains useful, detailed information about each video in the list of top 2500 most viewed Youtube.
- These are songs that publish on both Youtube and Spotify, which each song can be found on both platforms. All of these lines have the same meaning as introduced above.
- However, among all data lines, there is a minority of lines that do not have all the data fields, due to many different subjective and objective reasons (such as data publishers' collection errors, the platform not also supports viewing video information, the information collected is not mandatory information so users do not provide it to the platform,...)

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #E7B10A; padding: 15px; font-size: 18px; font-weight: bold; border-radius: 0 0 0 0">
🧮🧮🧮  In our data, each line contains data of a music song on both <font color=#E64848>youtube</font> and <font color=#3CCF4E>spotify</font> platform. 
</div>

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  3. Does the raw data have duplicate rows? 🧮
</div>

🚨  We need to check for the duplicated rows since duplicated rows don't have much meaning, even affect the overal insight of data.\
🚨  Therefore, each line should differ from each others.\
🚨  Having duplicated rows could lead to wrong insights for questions required counting.\
🚨  In this part, we will delve into check for duplicated rows. If duplicated rows exist, then drop it.

In [7]:
index = raw_df.index
detectDupSeries = index.duplicated(keep='first')
num_duplicated_rows = detectDupSeries.sum()

In [8]:
print(f'Number of duplicated rows: {num_duplicated_rows}')

Number of duplicated rows: 0


### 🎊🎊🎊 **Wonderful! Since our data doesn't contain duplicate rows, we don't need to delete any rows.** 🎊🎊🎊

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  4. What does each column mean?? 🧮
</div>

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #E7B10A; padding-left: 25px; font-size: 20px; font-weight: bold; border-radius: 0 0 0 0">
  4.1. What does each column mean ?
</div>

<table style="width: 100%; border-collapse: collapse;font-family: 'Trebuchet MS', Arial, sans-serif;">
  <thead>
    <tr>
      <th style="min-width: 10px; width: 30px; background-color: #4E70BB; color: white; font-size :20px; font-weight: bold; text-align: left; padding: 8px;">No</th>
      <th style="background-color: #4E70BB; color: white; font-size :20px; font-weight: bold; text-align: left; padding: 8px;">Column Name</th>
      <th style="background-color: #4E70BB; color: white; font-size :20px; font-weight: bold; text-align: left; padding: 8px;">Meaning</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #FFAA33; text-align: left;">1</td>
      <td style="font-weight: bold; font-size: 17px; color: #FFAA33; text-align: left;">Id</td>
      <td style="font-size: 17px; text-align: left;">Id code for a particular video on Youtube.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #FF0077; text-align: left;">2</td>
      <td style="font-weight: bold; font-size: 17px; color: #FF0077; text-align: left;">View</td>
      <td style="font-size: 17px; text-align: left;">Total view for a video on Youtube.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #009A80; text-align: left;">3</td>
      <td style="font-weight: bold; font-size: 17px; color: #009A80; text-align: left;">Like</td>
      <td style="font-size: 17px; text-align: left;">Number of likes for a video on Youtube.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #7DCE13; text-align: left;">4</td>
      <td style="font-weight: bold; font-size: 17px; color: #7DCE13; text-align: left;">Duration</td>
      <td style="font-size: 17px; text-align: left;">The length of a video on Youtube.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #EAE509; text-align: left;">5</td>
      <td style="font-weight: bold; font-size: 17px; color: #EAE509; text-align: left;">Channel_name</td>
      <td style="font-size: 17px; text-align: left;">Name of the channel uploading the video on Youtube.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #3081D0; text-align: left;">6</td>
      <td style="font-weight: bold; font-size: 17px; color: #3081D0; text-align: left;">Subscriber</td>
      <td style="font-size: 17px; text-align: left;">Number of subscribers of the channel on Youtube.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #FFAA33; text-align: left;">7</td>
      <td style="font-weight: bold; font-size: 17px; color: #FFAA33; text-align: left;">Country</td>
      <td style="font-size: 17px; text-align: left;">The territory (a nation) where the video is originally uploaded on Youtube.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #FF0077; text-align: left;">8</td>
      <td style="font-weight: bold; font-size: 17px; color: #FF0077; text-align: left;">Pubish_time</td>
      <td style="font-size: 17px; text-align: left;">The publish time of a video on Youtube.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #00A9FF; text-align: left;">9</td>
      <td style="font-weight: bold; font-size: 17px; color: #00A9FF; text-align: left;">Hashtag</td>
      <td style="font-size: 17px; text-align: left;">The number of hashtags on the description part of a video on Youtube.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #7DCE13; text-align: left;">10</td>
      <td style="font-weight: bold; font-size: 17px; color: #7DCE13; text-align: left;">Most view per day</td>
      <td style="font-size: 17px; text-align: left;">The highest number of views that is recorded in a day on Youtube.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #EAE509; text-align: left;">11</td>
      <td style="font-weight: bold; font-size: 17px; color: #EAE509; text-align: left;">Most-view-date</td>
      <td style="font-size: 17px; text-align: left;">The date that video on Youtube receive the highest number of views.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #3081D0; text-align: left;">12</td>
      <td style="font-weight: bold; font-size: 17px; color: #3081D0; text-align: left;">Highest rank</td>
      <td style="font-size: 17px; text-align: left;">The highest rank of video that have been recorded on Youtube since the publish date.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #FFAA33; text-align: left;">13</td>
      <td style="font-weight: bold; font-size: 17px; color: #FFAA33; text-align: left;">Time to Highest rank</td>
      <td style="font-size: 17px; text-align: left;">The duration from the publish date to the date achieving highest rank.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #FF0077; text-align: left;">14</td>
      <td style="font-weight: bold; font-size: 17px; color: #FF0077; text-align: left;">Charted-duration</td>
      <td style="font-size: 17px; text-align: left;">The duration that a song still sit in the chart.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #009A80; text-align: left;">15</td>
      <td style="font-weight: bold; font-size: 17px; color: #009A80; text-align: left;">Song name</td>
      <td style="font-size: 17px; text-align: left;">Name of the song on record.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #7DCE13; text-align: left;">16</td>
      <td style="font-weight: bold; font-size: 17px; color: #7DCE13; text-align: left;">Artist name</td>
      <td style="font-size: 17px; text-align: left;">Artist produced the song.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #EAE509; text-align: left;">17</td>
      <td style="font-weight: bold; font-size: 17px; color: #EAE509; text-align: left;">Release Date (Spotify)</td>
      <td style="font-size: 17px; text-align: left;">Release date of the song on Spotify platform.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #3081D0; text-align: left;">18</td>
      <td style="font-weight: bold; font-size: 17px; color: #3081D0; text-align: left;">Popularity Score (Spotify)</td>
      <td style="font-size: 17px; text-align: left;">Score of popularity on Spotify on scale of 100.</td>
    </tr>
    <tr>
      <td style="font-weight: bold; font-size: 17px; color: #FFAA33; text-align: left;">19</td>
      <td style="font-weight: bold; font-size: 17px; color: #FFAA33; text-align: left;">Genre</td>
      <td style="font-size: 17px; text-align: left;">Genre of the artist who produce the song.</td>
    </tr>
  </tbody>
</table>

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #E7B10A; padding-left: 25px; font-size: 20px; font-weight: bold; border-radius: 0 0 0 0">
  4.2. Comment and Solve.
</div>

🔱 From the table above, we observe that there exist two columns names `Song name` and `Artist name` that cover the meaning of the column `Title`.\
🔱 Hence, we decide to drop to column `Title` before moving to process other parts.

In [9]:
raw_df = raw_df.drop('Title', axis=1)
raw_df.head(10)

Id          View        Like Duration       Channel_name  \
0  9bZkp7q19f0  4.977100e+09  27830052.0  0:04:13        officialpsy   
1  hT_nvWreIhg  3.927344e+09  17491157.0  0:04:44    OneRepublicVEVO   
2  JGwWNGJdvx8  6.149237e+09  32323841.0  0:04:24         Ed Sheeran   
3  lp-EO5I60KA  3.700048e+09  14912926.0  0:04:57         Ed Sheeran   
4  CevxZvSJLk8  3.918655e+09  16596026.0  0:04:30      KatyPerryVEVO   
5  kJQP7kiw5Fk  8.327580e+09  52536744.0  0:04:42      LuisFonsiVEVO   
6  FzG4uDgje3M  4.494597e+09  18262030.0  0:02:33      Ultra Records   
7  pRpeEdMmmQ0  3.795632e+09  21936733.0  0:03:31        shakiraVEVO   
8  fRh_vgS2dFE  3.741592e+09  16306887.0  0:03:26   JustinBieberVEVO   
9  RgKAFK5djSk  6.108823e+09  42262511.0  0:03:58  Wiz Khalifa Music   

   Subscriber Country         Publish_time  Hasgtag  Most view per day  \
0  18400000.0     NaN  2012-07-15 07:46:32      4.0           14924298   
1   5470000.0     NaN  2013-05-31 07:00:36      2.0            3288973   
2  53900000.0     NaN  2017-01-30 10:57:50      3.0           14390704   
3  53900000.0     NaN  2014-10-07 13:57:37      3.0            3771622   
4  24600000.0      US  2013-09-05 20:00:22      0.0           11294380   
5  16900000.0      US  2017-01-13 05:00:02      8.0           25794523   
6  29900000.0     NaN  2018-04-05 19:51:33      0.0           14419156   
7  18400000.0     NaN  2010-06-04 22:30:35      4.0            3551141   
8  31500000.0      US  2015-10-22 20:00:02      5.0           11051088   
9     15200.0     NaN  2015-04-07 03:00:03      1.0            8818084   

  Most-view-date  Highest rank  Time to Highest rank  Charted-duration  \
0     2012/12/21           1.0                  36.0             482.0   
1     2018/11/10           4.0                   NaN             482.0   
2     2017/05/13           1.0                   4.0             356.0   
3     2015/02/14           3.0                   1.0             304.0   
4     2013/09/07           2.0                   5.0             445.0   
5     2017/08/05           1.0                  35.0             359.0   
6     2018/04/22           1.0                   5.0             296.0   
7     2010/06/26           1.0                   8.0             622.0   
8     2015/12/31           1.0                  15.0             115.0   
9     2015/05/23           1.0                  17.0             449.0   

                                           Song Name    Artist Name  \
0                              Gangnam Style (강남스타일)            PSY   
1                                     Counting Stars    OneRepublic   
2                                       Shape of You     Ed Sheeran   
3                                  Thinking out Loud     Ed Sheeran   
4                                               Roar     Katy Perry   
5                                          Despacito     Luis Fonsi   
6                 Dame Tu Cosita (feat. Cutty Ranks)      El Chombo   
7  Waka Waka (This Time for Africa) [The Official...        Shakira   
8                                              Sorry  Justin Bieber   
9                 See You Again (feat. Charlie Puth)    Wiz Khalifa   

  Release Date (Spotify)  Popularity Score (Spotify)  \
0             2012-01-01                          75   
1             2013-01-01                          88   
2             2017-03-03                          88   
3             2014-06-21                          85   
4             2013-10-18                          74   
5             2019-02-01                          81   
6             2018-05-10                          39   
7             2010-05-28                          69   
8             2015-11-13                          85   
9             2016-01-29                          61   

                                               Genre  
0             ['k-rap', 'korean old school hip hop']  
1                                 ['panamanian pop']  
2  ['colombian pop', 'dan

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding-left: 20px; padding-bottom:10px;padding-top:10px;font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  5.  What data type does each column currently have? 🧮
</div>
<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding-left: 20px; padding-bottom:20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
      Are there any columns whose data types are not suitable for further processing? 🧮
</div>

👀 Next we will take a look at the data type of each column.

In [10]:
dtypes = raw_df.dtypes

In [11]:
dtypes

Id                             object
View                          float64
Like                          float64
Duration                       object
Channel_name                   object
Subscriber                    float64
Country                        object
Publish_time                   object
Hasgtag                       float64
Most view per day               int64
Most-view-date                 object
Highest rank                  float64
Time to Highest rank          float64
Charted-duration              float64
Song Name                      object
Artist Name                    object
Release Date (Spotify)         object
Popularity Score (Spotify)      int64
Genre                          object
dtype: object

👀 `Most-view-date`, `Publish_time` and `Release Date (Spotify)` are problaly date-time data type.\
👀 We convert these 2 columns into ***date-time***. 

In [12]:
Most_view_date = []
Publish_time = []
Release_date = []

date_format = ['%Y', '%Y-%m', '%Y-%m-%d']

for date, time, release_date in zip(raw_df['Most-view-date'], raw_df['Publish_time'], raw_df['Release Date (Spotify)']):
    date = np.nan if pd.isna(date) else datetime.strptime(date, '%Y/%m/%d')
    time = np.nan if pd.isna(time) else datetime.strptime(time, '%Y-%m-%d %H:%M:%S')
    if (pd.isna(release_date)):
        release_date = np.nan
    elif ('-' not in (release_date)):
        release_date = datetime.strptime(release_date, '%Y')
    elif (len(release_date) == 7):
        release_date = datetime.strptime(release_date, '%Y-%m')
    else:
        release_date = datetime.strptime(release_date, '%Y-%m-%d')
    
    Most_view_date.append(date)
    Publish_time.append(time)
    Release_date.append(release_date)
    
raw_df['Most-view-date'] = Most_view_date
raw_df['Publish_time'] = Publish_time
raw_df['Release Date (Spotify)'] = Release_date

👀 `View`, `Like`, `Subscriber`, `Hashtag`, `Highest rank`, `Time to Highest rank` and `Charted-duration` can be observes as integer data type.\
👀 We convert these columns into ***integer***.

In [13]:
def convert_to_int(value):
    try:
        return int(value)
    except ValueError:
        return None

raw_df['View'] = raw_df['View'].apply(convert_to_int)
raw_df['Like'] = raw_df['Like'].apply(convert_to_int)
raw_df['Subscriber'] = raw_df['Subscriber'].apply(convert_to_int)
raw_df['Hasgtag'] = raw_df['Hasgtag'].apply(convert_to_int)
raw_df['Highest rank'] = raw_df['Highest rank'].apply(convert_to_int)
raw_df['Time to Highest rank'] = raw_df['Time to Highest rank'].apply(convert_to_int)
raw_df['Charted-duration'] = raw_df['Charted-duration'].apply(convert_to_int)

👀 Now let's take a look again the data types of each column.

In [14]:
raw_df.dtypes

Id                                    object
View                                 float64
Like                                 float64
Duration                              object
Channel_name                          object
Subscriber                           float64
Country                               object
Publish_time                  datetime64[ns]
Hasgtag                              float64
Most view per day                      int64
Most-view-date                datetime64[ns]
Highest rank                         float64
Time to Highest rank                 float64
Charted-duration                     float64
Song Name                             object
Artist Name                           object
Release Date (Spotify)        datetime64[ns]
Popularity Score (Spotify)             int64
Genre                                 object
dtype: object

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #FFC436; padding-left: 15px; font-size: 18px; font-weight: bold; border-radius: 0 0 0 0">
We can see that although we try to convert the values of these columns into integer, their data types are still float64. 
</div>
<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #FFC436; padding: 15px; font-size: 18px; font-weight: bold; border-radius: 0 0 0 0">
This is beacause integer data type doesn't allow for <font color=#E64848>NaN</font> so if there is a <font color=#E64848>NaN</font> value in the column, the column data type will automatically upcasted to <font color=#3CCF4E style="font-weight: bold">float</font>.
</div>

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  6. For each column with numeric data type, how are the values distributed ? 🧮
</div> 

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #E7B10A; padding-left: 25px; font-size: 20px; font-weight: bold; border-radius: 0 0 0 0">
  6.1. Calculating  Descriptive Statistics?
</div>

❄️  For columns with numeric data types, we will calculate:   
- `Percentage` (from 0 to 100) of missing values
- The `min`
- The `lower quartile` (phân vị 25)
- The `median` (phân vị 50)
- The `upper quartile` (phân vị 75)
- The `max`

❄️  First we need to select columns with numeric data type.

In [15]:
num_col_info_df = raw_df.select_dtypes(include='number')

In [16]:
def missing_ratio(s):
    return (s.isna().mean() * 100).round(1)

def median(df):
    return (df.quantile(0.5)).round(1)

def lower_quartile(df):
    return (df.quantile(0.25)).round(1)

def upper_quartile(df):
    return (df.quantile(0.75)).round(1)

In [17]:
num_col_info_df = num_col_info_df.agg([missing_ratio, "min", lower_quartile, median, upper_quartile, "max"])
num_col_info_df

View        Like   Subscriber  Hasgtag  \
missing_ratio   4.000000e-01         0.5          0.4      0.4   
min             3.244410e+08    471866.0       7330.0      0.0   
lower_quartile  3.982876e+08   2089275.2    2935000.0      0.0   
median          5.150177e+08   3129072.0    9160000.0      3.0   
upper_quartile  7.857353e+08   5063792.2   24600000.0      4.0   
max             8.327580e+09  52536744.0  254000000.0     81.0   

                Most view per day  Highest rank  Time to Highest rank  \
missing_ratio                 0.0          25.0                  87.2   
min                      141350.0           1.0                   1.0   
lower_quartile           765111.5           6.0                   1.0   
median                  1837817.5          21.0                   1.0   
upper_quartile          4233410.5          58.0                   3.0   
max                    88851354.0         200.0                  36.0   

                Charted-duration  Popularity Score (Spotify)  
missing_ratio               25.0                         0.0  
min                          1.0                         0.0  
lower_quartile              13.0                        59.0  
median                      25.0                        71.0  
upper_quartile              47.0                        78.0  
max                        622.0                        97.0

📈 `COMMENT:`\
❄️  After identifying the basic statistical numbers that describe your data, we further need to determine the features that have a large number of missing values. Such features are not useful for the analysis stage and must be removed from the dataset.   
❄️  Usually, if the percentage of missing values is greater than 75%, the column is dropped from the dataframe and an updated dataframe is returned.

In [18]:
def drop_missing_features(df: pd.DataFrame, missing_lst: dict = dict(num_col_info_df.iloc[0]), threshold: float = 75.0) -> pd.DataFrame:
    if (df is None) or (missing_lst is None) or (threshold is None):
        print(f"[LOG] Invalid.")
        raise ValueError
    
    # Create a copy of the dataframe to avoid changing the original
    df_cp = df.copy()
    
    # List to maintain the columns to drop
    cols_to_trim = []
    
    for key, value in missing_lst.items():
        if float(value) > threshold:
            cols_to_trim.append(key)
            
    if len(cols_to_trim) > 0:
        # drop the columns identified using the dataframe drop() method
        df_cp = df_cp.drop(columns=cols_to_trim)
        print("[LOG] Dropped column(s): " + " ".join(cols_to_trim))
    else:
        print("[LOG] Have no column(s) to trim.")
        
    return df_cp

In [19]:
raw_df = drop_missing_features(raw_df, dict(num_col_info_df.iloc[0]))

[LOG] Dropped column(s): Time to Highest rank


In [20]:
raw_df.head()

Id          View        Like Duration     Channel_name  \
0  9bZkp7q19f0  4.977100e+09  27830052.0  0:04:13      officialpsy   
1  hT_nvWreIhg  3.927344e+09  17491157.0  0:04:44  OneRepublicVEVO   
2  JGwWNGJdvx8  6.149237e+09  32323841.0  0:04:24       Ed Sheeran   
3  lp-EO5I60KA  3.700048e+09  14912926.0  0:04:57       Ed Sheeran   
4  CevxZvSJLk8  3.918655e+09  16596026.0  0:04:30    KatyPerryVEVO   

   Subscriber Country        Publish_time  Hasgtag  Most view per day  \
0  18400000.0     NaN 2012-07-15 07:46:32      4.0           14924298   
1   5470000.0     NaN 2013-05-31 07:00:36      2.0            3288973   
2  53900000.0     NaN 2017-01-30 10:57:50      3.0           14390704   
3  53900000.0     NaN 2014-10-07 13:57:37      3.0            3771622   
4  24600000.0      US 2013-09-05 20:00:22      0.0           11294380   

  Most-view-date  Highest rank  Charted-duration              Song Name  \
0     2012-12-21           1.0             482.0  Gangnam Style (강남스타일)   
1     2018-11-10           4.0             482.0         Counting Stars   
2     2017-05-13           1.0             356.0           Shape of You   
3     2015-02-14           3.0             304.0      Thinking out Loud   
4     2013-09-07           2.0             445.0                   Roar   

   Artist Name Release Date (Spotify)  Popularity Score (Spotify)  \
0          PSY             2012-01-01                          75   
1  OneRepublic             2013-01-01                          88   
2   Ed Sheeran             2017-03-03                          88   
3   Ed Sheeran             2014-06-21                          85   
4   Katy Perry             2013-10-18                          74   

                                               Genre  
0             ['k-rap', 'korean old school hip hop']  
1                                 ['panamanian pop']  
2  ['colombian pop', 'dance pop', 'latin pop', 'p...  
3                  ['latin pop', 'puerto rican pop']  
4                            ['canadian pop', 'pop']

📈`COMMENT:`\
❄️  After remove features which have large missing values, our dataframe still have missing values. So that, we need to fill these missing values so that they can be used in analysis.  
❄️  There's many ways to fill in the missing values. Here we will use the median value of each column to fill in.

In [21]:
def filling_missing_value(df: pd.DataFrame) -> pd.DataFrame:
    df = df.replace({None: np.nan})
    
    if (df is None):
        print(f"[LOG] Invalid.")
        raise ValueError
    
    # Create a copy of the dataframe to avoid changing the original
    df_cp = df.copy()
    
    # Get the list of columns in the dataframe
    cols_list = list(df_cp.columns)
    
    # Exclude the last column - Country, Year
    cols_list.remove("Id")
    cols_list.remove("Duration")
    cols_list.remove("Channel_name")
    cols_list.remove("Publish_time")
    cols_list.remove("Country")
    cols_list.remove("Most-view-date")
    cols_list.remove("Song Name")
    cols_list.remove("Artist Name")
    cols_list.remove("Release Date (Spotify)")
    cols_list.remove("Genre")
    
    # Replace all None values with NaN, fillna only works on nans
    df_cp.fillna(value=np.nan, inplace=True)
    
    # Replace all NaN values with the mean of the column values
    for col in cols_list:
        df_cp[col].fillna((df_cp[col].median()), inplace=True)
        
    return df_cp

In [22]:
raw_df = filling_missing_value(df=raw_df)

In [23]:
raw_df.head()

Id          View        Like Duration     Channel_name  \
0  9bZkp7q19f0  4.977100e+09  27830052.0  0:04:13      officialpsy   
1  hT_nvWreIhg  3.927344e+09  17491157.0  0:04:44  OneRepublicVEVO   
2  JGwWNGJdvx8  6.149237e+09  32323841.0  0:04:24       Ed Sheeran   
3  lp-EO5I60KA  3.700048e+09  14912926.0  0:04:57       Ed Sheeran   
4  CevxZvSJLk8  3.918655e+09  16596026.0  0:04:30    KatyPerryVEVO   

   Subscriber Country        Publish_time  Hasgtag  Most view per day  \
0  18400000.0     NaN 2012-07-15 07:46:32      4.0           14924298   
1   5470000.0     NaN 2013-05-31 07:00:36      2.0            3288973   
2  53900000.0     NaN 2017-01-30 10:57:50      3.0           14390704   
3  53900000.0     NaN 2014-10-07 13:57:37      3.0            3771622   
4  24600000.0      US 2013-09-05 20:00:22      0.0           11294380   

  Most-view-date  Highest rank  Charted-duration              Song Name  \
0     2012-12-21           1.0             482.0  Gangnam Style (강남스타일)   
1     2018-11-10           4.0             482.0         Counting Stars   
2     2017-05-13           1.0             356.0           Shape of You   
3     2015-02-14           3.0             304.0      Thinking out Loud   
4     2013-09-07           2.0             445.0                   Roar   

   Artist Name Release Date (Spotify)  Popularity Score (Spotify)  \
0          PSY             2012-01-01                          75   
1  OneRepublic             2013-01-01                          88   
2   Ed Sheeran             2017-03-03                          88   
3   Ed Sheeran             2014-06-21                          85   
4   Katy Perry             2013-10-18                          74   

                                               Genre  
0             ['k-rap', 'korean old school hip hop']  
1                                 ['panamanian pop']  
2  ['colombian pop', 'dance pop', 'latin pop', 'p...  
3                  ['latin pop', 'puerto rican pop']  
4                            ['canadian pop', 'pop']

In [24]:
raw_df.select_dtypes(include='number').agg([missing_ratio, "min", lower_quartile, median, upper_quartile, "max"])

View        Like   Subscriber  Hasgtag  \
missing_ratio   0.000000e+00         0.0          0.0      0.0   
min             3.244410e+08    471866.0       7330.0      0.0   
lower_quartile  3.985463e+08   2094270.2    2950000.0      0.0   
median          5.150177e+08   3129072.0    9160000.0      3.0   
upper_quartile  7.838052e+08   5046276.8   24600000.0      4.0   
max             8.327580e+09  52536744.0  254000000.0     81.0   

                Most view per day  Highest rank  Charted-duration  \
missing_ratio                 0.0           0.0               0.0   
min                      141350.0           1.0               1.0   
lower_quartile           765111.5           9.8              17.0   
median                  1837817.5          21.0              25.0   
upper_quartile          4233410.5          44.0              37.0   
max                    88851354.0         200.0             622.0   

                Popularity Score (Spotify)  
missing_ratio                          0.0  
min                                    0.0  
lower_quartile                        59.0  
median                                71.0  
upper_quartile                        78.0  
max                                   97.0

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  7. For each column with a non-numeric data type, how are the values distributed ? 🧮
</div> 

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #E7B10A; padding-left: 25px; font-size: 20px; font-weight: bold; border-radius: 0 0 0 0">
  7.1. Calculating  Descriptive Statistics: Missing ratio, Number of values and Value ratios.
</div>

🌸 For columns with non-numeric data types, we calculate:

✨ Percentage (from 0 to 100) of missing values.\
✨ Number of values (the values here are different values and we do not consider missing values): with columns whose type is categorical, it is a set with a finite number of categories. Directly counting the number of values in these columns doesn't make much sense, so it's better to count the number of elements of all types.\
(Số lượng các giá trị (các giá trị ở đây là các giá trị khác nhau và ta không xét giá trị thiếu): với cột mà có kiểu là categorical, nó là một tập hợp có số lượng hữu hạn các loại. Việc đếm trực tiếp số lượng các giá trị trong những cột này không có nhiều ý nghĩa, nên tốt hơn hết là mà sẽ đếm số lượng phần tử các loại).\
✨ The percentage (from 0 to 100) of each value is sorted by decreasing percentage (we do not consider missing values, the ratio is the ratio compared to the number of non-missing values): you use a dictionary to store , key is the value, value is the percentage; With the column corresponding to each type, the method is similar to above.

In [25]:
cat_col_info_df = raw_df.select_dtypes(include='object')

In [26]:
def missing_ratio(s):
    return (s.isna().mean() * 100).round(1)

def num_values(s):
    s = s.str.split(';')
    s = s.explode()
    return len(s.value_counts())

def value_ratios(s):
    s = s.str.split(';')
    s = s.explode()
    totalCount = (~s.isna()).sum()
    return ((s.value_counts()/totalCount*100).round(1)).to_dict()

In [27]:
cat_col_info_df = cat_col_info_df.agg([missing_ratio, num_values, value_ratios])
cat_col_info_df

Id  \
missing_ratio                                                0.0   
num_values                                                  2500   
value_ratios   {'9bZkp7q19f0': 0.0, '_EyZUTDAH0U': 0.0, 'f5aD...   

                                                        Duration  \
missing_ratio                                                0.4   
num_values                                                   322   
value_ratios   {'0:03:54': 1.3, '0:03:38': 1.2, '0:03:55': 1....   

                                                    Channel_name  \
missing_ratio                                                0.4   
num_values                                                   893   
value_ratios   {'T-Series': 4.8, 'Zee Music Company': 1.0, 'B...   

                                                         Country  \
missing_ratio                                               44.6   
num_values                                                    40   
value_ratios   {'US': 38.4, 'IN': 21.7, 'BR': 8.7, 'PR': 6.4,...   

                                                       Song Name  \
missing_ratio                                                0.0   
num_values                                                  2359   
value_ratios   {'Pepas': 0.2, 'I Am Your Gummy Bear (The Gumm...   

                                                     Artist Name  \
missing_ratio                                                0.0   
num_values                                                  1083   
value_ratios   {'Bad Bunny': 1.0, 'Justin Bieber': 0.9, 'BTS'...   

                                                           Genre  
missing_ratio                                                0.0  
num_values                                                   716  
value_ratios   {'[]': 4.4, '['pop']': 4.2, '['filmi', 'modern...

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #E7B10A; padding-left: 25px; font-size: 20px; font-weight: bold; border-radius: 0 0 0 0">
  7.2. Process data on Genre column
</div>

👀 Let's take a look on the column named `Genre` of the dataframe.

In [28]:
raw_df['Genre']

0                  ['k-rap', 'korean old school hip hop']
1                                      ['panamanian pop']
2       ['colombian pop', 'dance pop', 'latin pop', 'p...
3                       ['latin pop', 'puerto rican pop']
4                                 ['canadian pop', 'pop']
                              ...                        
2495                 ['desi pop', 'filmi', 'punjabi pop']
2496                               ['modern country pop']
2497    ['agronejo', 'arrocha', 'sertanejo', 'sertanej...
2498                              ['canadian pop', 'pop']
2499    ['sertanejo', 'sertanejo pop', 'sertanejo univ...
Name: Genre, Length: 2500, dtype: object

📜 `OBSERVATION:`

- The string of the column `Genre` seems to be so complicated with the appearance of symbol like `[]`, `''`.
- Hence, let's drop it for better visualiztion.

In [29]:
raw_df['Genre'] = raw_df['Genre'].str.replace('[', '')
raw_df['Genre'] = raw_df['Genre'].str.replace(']', '')
raw_df['Genre'] = raw_df['Genre'].str.replace('\'', '')

In [30]:
raw_df['Genre']

0                        k-rap, korean old school hip hop
1                                          panamanian pop
2                colombian pop, dance pop, latin pop, pop
3                             latin pop, puerto rican pop
4                                       canadian pop, pop
                              ...                        
2495                         desi pop, filmi, punjabi pop
2496                                   modern country pop
2497    agronejo, arrocha, sertanejo, sertanejo univer...
2498                                    canadian pop, pop
2499    sertanejo, sertanejo pop, sertanejo universitario
Name: Genre, Length: 2500, dtype: object

🎊🎊🎊 The string looks better now !

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  8. Rearrange the order of column feature 🧮
</div> 

👀 Let's take a look on the whole dataframe again.

In [31]:
raw_df.head()

Id          View        Like Duration     Channel_name  \
0  9bZkp7q19f0  4.977100e+09  27830052.0  0:04:13      officialpsy   
1  hT_nvWreIhg  3.927344e+09  17491157.0  0:04:44  OneRepublicVEVO   
2  JGwWNGJdvx8  6.149237e+09  32323841.0  0:04:24       Ed Sheeran   
3  lp-EO5I60KA  3.700048e+09  14912926.0  0:04:57       Ed Sheeran   
4  CevxZvSJLk8  3.918655e+09  16596026.0  0:04:30    KatyPerryVEVO   

   Subscriber Country        Publish_time  Hasgtag  Most view per day  \
0  18400000.0     NaN 2012-07-15 07:46:32      4.0           14924298   
1   5470000.0     NaN 2013-05-31 07:00:36      2.0            3288973   
2  53900000.0     NaN 2017-01-30 10:57:50      3.0           14390704   
3  53900000.0     NaN 2014-10-07 13:57:37      3.0            3771622   
4  24600000.0      US 2013-09-05 20:00:22      0.0           11294380   

  Most-view-date  Highest rank  Charted-duration              Song Name  \
0     2012-12-21           1.0             482.0  Gangnam Style (강남스타일)   
1     2018-11-10           4.0             482.0         Counting Stars   
2     2017-05-13           1.0             356.0           Shape of You   
3     2015-02-14           3.0             304.0      Thinking out Loud   
4     2013-09-07           2.0             445.0                   Roar   

   Artist Name Release Date (Spotify)  Popularity Score (Spotify)  \
0          PSY             2012-01-01                          75   
1  OneRepublic             2013-01-01                          88   
2   Ed Sheeran             2017-03-03                          88   
3   Ed Sheeran             2014-06-21                          85   
4   Katy Perry             2013-10-18                          74   

                                      Genre  
0          k-rap, korean old school hip hop  
1                            panamanian pop  
2  colombian pop, dance pop, latin pop, pop  
3               latin pop, puerto rican pop  
4                         canadian pop, pop

📈 `COMMENT:`

🌵 It looks quite weird that the `Song Name` column, `Artist Name` column and `Channel_name` located in the middle of the dataframe. So we will move them to the head of dataframe, behind the column `Id`.\
🌵 Let's bring all the numerical column to the head part of the dataframe, and non-numeric column to the tail of the dataframe for better visualization.

In [32]:
#rearrange the position of columns
raw_df = raw_df[['Id', 'Song Name', 'Artist Name', 'Channel_name', 'View', 'Like', 'Subscriber', 'Hasgtag', 'Most view per day', 'Highest rank', 'Most-view-date','Charted-duration', 'Popularity Score (Spotify)', 'Duration', 'Country', 'Publish_time', 'Release Date (Spotify)', 'Genre']]
raw_df.head()

Id              Song Name  Artist Name     Channel_name  \
0  9bZkp7q19f0  Gangnam Style (강남스타일)          PSY      officialpsy   
1  hT_nvWreIhg         Counting Stars  OneRepublic  OneRepublicVEVO   
2  JGwWNGJdvx8           Shape of You   Ed Sheeran       Ed Sheeran   
3  lp-EO5I60KA      Thinking out Loud   Ed Sheeran       Ed Sheeran   
4  CevxZvSJLk8                   Roar   Katy Perry    KatyPerryVEVO   

           View        Like  Subscriber  Hasgtag  Most view per day  \
0  4.977100e+09  27830052.0  18400000.0      4.0           14924298   
1  3.927344e+09  17491157.0   5470000.0      2.0            3288973   
2  6.149237e+09  32323841.0  53900000.0      3.0           14390704   
3  3.700048e+09  14912926.0  53900000.0      3.0            3771622   
4  3.918655e+09  16596026.0  24600000.0      0.0           11294380   

   Highest rank Most-view-date  Charted-duration  Popularity Score (Spotify)  \
0           1.0     2012-12-21             482.0                          75   
1           4.0     2018-11-10             482.0                          88   
2           1.0     2017-05-13             356.0                          88   
3           3.0     2015-02-14             304.0                          85   
4           2.0     2013-09-07             445.0                          74   

  Duration Country        Publish_time Release Date (Spotify)  \
0  0:04:13     NaN 2012-07-15 07:46:32             2012-01-01   
1  0:04:44     NaN 2013-05-31 07:00:36             2013-01-01   
2  0:04:24     NaN 2017-01-30 10:57:50             2017-03-03   
3  0:04:57     NaN 2014-10-07 13:57:37             2014-06-21   
4  0:04:30      US 2013-09-05 20:00:22             2013-10-18   

                                      Genre  
0          k-rap, korean old school hip hop  
1                            panamanian pop  
2  colombian pop, dance pop, latin pop, pop  
3               latin pop, puerto rican pop  
4                         canadian pop, pop

👀 The name of column featured `Hashtag` seemed to be misspelt.\
👀 Let's correct it.

In [33]:
#change name of column Hasgtag to Hashtag
raw_df = raw_df.rename(columns={'Hasgtag': 'Hashtag'})
raw_df.head()

Id              Song Name  Artist Name     Channel_name  \
0  9bZkp7q19f0  Gangnam Style (강남스타일)          PSY      officialpsy   
1  hT_nvWreIhg         Counting Stars  OneRepublic  OneRepublicVEVO   
2  JGwWNGJdvx8           Shape of You   Ed Sheeran       Ed Sheeran   
3  lp-EO5I60KA      Thinking out Loud   Ed Sheeran       Ed Sheeran   
4  CevxZvSJLk8                   Roar   Katy Perry    KatyPerryVEVO   

           View        Like  Subscriber  Hashtag  Most view per day  \
0  4.977100e+09  27830052.0  18400000.0      4.0           14924298   
1  3.927344e+09  17491157.0   5470000.0      2.0            3288973   
2  6.149237e+09  32323841.0  53900000.0      3.0           14390704   
3  3.700048e+09  14912926.0  53900000.0      3.0            3771622   
4  3.918655e+09  16596026.0  24600000.0      0.0           11294380   

   Highest rank Most-view-date  Charted-duration  Popularity Score (Spotify)  \
0           1.0     2012-12-21             482.0                          75   
1           4.0     2018-11-10             482.0                          88   
2           1.0     2017-05-13             356.0                          88   
3           3.0     2015-02-14             304.0                          85   
4           2.0     2013-09-07             445.0                          74   

  Duration Country        Publish_time Release Date (Spotify)  \
0  0:04:13     NaN 2012-07-15 07:46:32             2012-01-01   
1  0:04:44     NaN 2013-05-31 07:00:36             2013-01-01   
2  0:04:24     NaN 2017-01-30 10:57:50             2017-03-03   
3  0:04:57     NaN 2014-10-07 13:57:37             2014-06-21   
4  0:04:30      US 2013-09-05 20:00:22             2013-10-18   

                                      Genre  
0          k-rap, korean old school hip hop  
1                            panamanian pop  
2  colombian pop, dance pop, latin pop, pop  
3               latin pop, puerto rican pop  
4                         canadian pop, pop

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  9. Is the collected data reasonable ?🧮
</div> 

This step make sure that collected data is reasonable, before moving on to the next step. These are some simple check: 
- `checking_view_vs_Like`: In reality, viewers need to access the video and increase views to be able to choose to like it (or not). Therefore, the number of views must always be greater than (or equal to) the number of likes, but absolutely cannot be smaller.
- `checking_view_vs_MostView`: A video every day will increase its views, but there cannot be a Most view per day that is greater than the total view (because the total view itself includes Most views per day and data from the previous and subsequent days).
- `checking_most_view_date`: All dates related to video statistics (such as the date with the highest views) must occur after the video's publication date.
- `checking_valid_PopularityScore`: All value in Popularity Score (Spotify) field must in range (0, 100)

In [7]:
def checking_view_vs_MostView(df):
    return len(df[df["View"] < df["Most view per day"]]) == 0

def checking_view_vs_Like(df):
    return len(df[df["View"] < df["Like"]]) == 0

def checking_most_view_date(df):
    return len(df[df["Publish_time"] >= df["Most-view-date"]]) == 0

def checking_valid_PopularityScore (df):
    return len(df[df["Popularity Score (Spotify)"] < 0]) == 0 and len(df[df["Popularity Score (Spotify)"] > 100]) == 0

`checking_view_vs_MostView` show that all rows have reasonable data by this criteria.

In [158]:
checking_view_vs_MostView(raw_df) == True

True

`checking_view_vs_Like` show that all rows have reasonable data by this criteria.

In [159]:
checking_view_vs_Like(raw_df) == True

True

`checking_most_view_date` show that all rows have reasonable data by this criteria.

In [160]:
checking_most_view_date(raw_df) == True

True

`checking_valid_PopularityScore` show that all rows have reasonable data by this criteria.

In [8]:
checking_valid_PopularityScore(raw_df) == True

True

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  10. Save processed data 🧮
</div> 

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #FFC436; padding-left: 15px; font-size: 18px; font-weight: bold; border-radius: 0 0 0 0">
Finally, we finish processing our data with our highest posibility.
</div>
<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #FFC436; padding-left: 15px;padding-top:10px; font-size: 18px; font-weight: bold; border-radius: 0 0 0 0">
We save the data sample with the name: <font color=#3CCF4E style="font-weight: bold">music_video_data</font> in folder processed.
</div>

`../data/processed/music_video_data.csv`

In [34]:
print(f"Total number of features: {raw_df.shape[1]}")
raw_df.dtypes

Total number of features: 18


Id                                    object
Song Name                             object
Artist Name                           object
Channel_name                          object
View                                 float64
Like                                 float64
Subscriber                           float64
Hashtag                              float64
Most view per day                      int64
Highest rank                         float64
Most-view-date                datetime64[ns]
Charted-duration                     float64
Popularity Score (Spotify)             int64
Duration                              object
Country                               object
Publish_time                  datetime64[ns]
Release Date (Spotify)        datetime64[ns]
Genre                                 object
dtype: object

In [35]:
raw_df.to_csv("../data/processed/music_video_data.csv", index=False)